### DISCUSSION TITLE
(`defer` API) - is there a way to transform `defer` data on the server instead of the client?

### DISCUSSION THREAD

**Discussion by cliffordfajardo** [remix-run/remix#4527] [Q&A] [answered]
Hi Remix community ⚛️ ,
Is there a way to manipulate/transform the data my non-critical `defer` data in my loader on the server instead of doing the manipulation on the client?

### Preface
When you're `await`-ing data in your loader, its easy to manipulate the response before it goes to the client like so:

```tsx
export async function loader({ request }: LoaderArgs) {
  const searchParams = new URL(request.url).searchParams;
  const people = await getPeople();
  const filteredPeople = filterPeople(people, searchParams);
  
return json({ people: filteredPeople });
}

export default function Index() {
  const data = useLoaderData<typeof loader>();

  return (
    <section>
      <h2> People Directory</h2>
      <ul>
        {data.people.map((person) => (
          <li key={person.key}>{JSON.stringify(person)}</li>
        ))}
      </ul>
    </section>
  );
}
```

However, when you are not using `await` its not possible to manipulate/filter the data in the loader since you're working with a promise, not an actual value:

```ts
export async function loader({ request }: LoaderArgs) {
  const people = getPeople();

  // Since `people` is a promise, I cant filter the data in here like in my `loader-example` example
  // ❌  people = filterPeople(people, searchParams);
  return defer({ people: people, });
}
```

<br/>
<br/>
This means I have to move the data manipulation logic for my `defer` data  into my component.
Though the example below works, I had to move logic that otherwise, perhaps, could have lived on the server into my component.

- 🌕import `useSearchParams` and use it in my component
- 🌕 filter my list inside of the `<Await>` children callback

```tsx
export async function loader({ request }: LoaderArgs) {
  const people = getPeople();
  return defer({ people: people, });
}

export default function Index() {
  const data = useLoaderData<typeof loader>();
  const [searchParams] = useSearchParams(); // 🌕

  return (
    <Suspense fallback={<h2>People directory loading...</h2>}>
      <Await
        resolve={data.people}
        errorElement={<div>Error</div>}
        children={(people) => {
          // 🌕  
          const filteredPeople = filterPeople(people || [], searchParams);
          return (
            <>
              <h2>People Directory</h2>
              <ul>
                {filteredPeople.map((person) => (
                  <li key={person.key}>{JSON.stringify(person)}</li>
                ))}
              </ul>
            </>
          );
        }}
      />
    </Suspense>
  );
}
```

### References
- stackblitz example of code above: https://stackblitz.com/edit/node-lrhrzm

**Comment by nakleiderer** [comment, marked as answer]
The defer api is built around promises. You can transform data by using Promise.then() or by wrapping your result in an async function. 

```js
export async function loader({ request }: LoaderArgs) {
  const searchParams = new URL(request.url).searchParams;
  const people = getPeople().then((people) =>  filterPeople(people || [], searchParams))

  return defer({
      people
    },
  });
}
```


```js

async function fetchFilteredPeople(searchParams) {
  const people = await getPeople()

  return filterPeople(people || [], searchParams)
}

export async function loader({ request }: LoaderArgs) {
  const searchParams = new URL(request.url).searchParams;
  const people = fetchFilteredPeople(searchParams)

  return defer({
      people
    },
  });
}
```

**Reply by cliffordfajardo** [reply]
Thanks @nakleiderer  - very clean solutions 
I'm embarrassed by how simple that seemingly was - you rock 💪 
 

 Updated the stackblitz code to include @nakleiderer  solutions. See the `defer-example2.tsx` file
 https://stackblitz.com/edit/node-lrhrzm?file=app%2Froutes%2Fdefer-example2.tsx
 
 
DEMO (using stackblitz code above ⬆️ )

https://user-images.githubusercontent.com/6743796/200941399-9d2c8749-11bb-4ee5-929c-49b309c58872.mp4

**Reply by kiliman** [reply]
Exactly, Remix is built on primitives both from the web as well as JavaScript. This simplifies things drastically as you don't need to learn a new API, but just use existing ones in novel ways.

Using Remix will make a better web developer as well as JavaScript developer.

**Comment by kevinreber** [comment]
@nakleiderer I forked @cliffordfajardo StackBlitz playground and upgraded all my Remix dependencies to the latest `v1.7.6`, but I'm receiving the following TypeScript errors:

```
Module '"@remix-run/node"' has no exported member 'defer'.
```
and
```
Module '"@remix-run/react"' has no exported member 'Await'.
```

Are `defer` and `Await` only usable on certain versions? I noticed @cliffordfajardo is using `deferred` for his Remix dependencies and saw this interesting roadmap for [Streaming/defer](https://github.com/remix-run/remix/issues/4486)

### References
- My StackBlitz playground: https://stackblitz.com/edit/node-srqtta?file=app%2Froutes%2Fdefer-example1.tsx,package.json

**Reply by machour** [reply]
Yep, you need to stay on the `deferred` branch for now
https://github.com/remix-run/remix/tree/deferred
